guardian/shortcuts.py

```diff
@@ -783,23 +785,36 @@ def get_objects_for_group(group, perms, klass=None, any_perm=False, accept_globa
         objects = queryset.filter(pk__in=pk_list)
         return objects
 
-    is_cast_integer = _is_cast_integer_pk(queryset)
-
     field_pk = fields[0]
     values = groups_obj_perms_queryset
 
-    if is_cast_integer:
+    cast_pk_field = _cast_pk_field(queryset)
+
+    if cast_pk_field:
         values = values.annotate(
-            obj_pk=Cast(field_pk, BigIntegerField())
+            obj_pk=Cast(field_pk, cast_pk_field())
         )
         field_pk = 'obj_pk'
 
     values = values.values_list(field_pk, flat=True)
     return queryset.filter(pk__in=values)
 
 
-def _is_cast_integer_pk(queryset):
-    return isinstance(queryset.model._meta.pk, (
-        IntegerField, AutoField, BigIntegerField,
-        PositiveIntegerField, PositiveSmallIntegerField,
-        SmallIntegerField))
+def _cast_pk_field(queryset):
+    if isinstance(
+        queryset.model._meta.pk,
+        (
+            IntegerField,
+            AutoField,
+            BigIntegerField,
+            PositiveIntegerField,
+            PositiveSmallIntegerField,
+            SmallIntegerField,
+        ),
+    ):
+        return BigIntegerField
+
+    if isinstance(queryset.model._meta.pk, UUIDField):
+        return UUIDField
```

Comment:
Related import: `from django.db import connection`

```suggestion
    if isinstance(queryset.model._meta.pk, UUIDField) and connection.features.has_native_uuid_field:
        return UUIDField
```